File: swift/Sources/AlgorithmsMultiverse/Searching.swift
// Searching Algorithms - Swift Implementation
// Comprehensive collection of search algorithms with iOS optimizations

import Foundation

/// Namespace for searching algorithms
public enum Searching {

    // MARK: - Binary Search Variants

    /// Standard binary search for sorted arrays
    /// - Parameters:
    ///   - array: Sorted array to search
    ///   - target: Element to find
    /// - Returns: Index of target or nil if not found
    /// - Complexity: O(log n) time, O(1) space
    public static func binarySearch<T: Comparable>(_ array: [T], target: T) -> Int? {
        var left = 0
        var right = array.count - 1

        while left <= right {
            let mid = left + (right - left) / 2

            if array[mid] == target {
                return mid
            } else if array[mid] < target {
                left = mid + 1
            } else {
                right = mid - 1
            }
        }

        return nil
    }

    /// Binary search with custom comparator
    public static func binarySearch<T>(
        _ array: [T],
        target: T,
        comparator: (T, T) -> ComparisonResult
    ) -> Int? {
        var left = 0
        var right = array.count - 1

        while left <= right {
            let mid = left + (right - left) / 2
            let result = comparator(array[mid], target)

            if result == .orderedSame {
                return mid
            } else if result == .orderedAscending {
                left = mid + 1
            } else {
                right = mid - 1
            }
        }

        return nil
    }

    /// Find first occurrence of target in sorted array
    /// - Complexity: O(log n) time, O(1) space
    public static func lowerBound<T: Comparable>(_ array: [T], target: T) -> Int {
        var left = 0
        var right = array.count

        while left < right {
            let mid = left + (right - left) / 2

            if array[mid] < target {
                left = mid + 1
            } else {
                right = mid
            }
        }

        return left
    }

    /// Find first position where target could be inserted
    /// - Complexity: O(log n) time, O(1) space
    public static func upperBound<T: Comparable>(_ array: [T], target: T) -> Int {
        var left = 0
        var right = array.count

        while left < right {
            let mid = left + (right - left) / 2

            if array[mid] <= target {
                left = mid + 1
            } else {
                right = mid
            }
        }

        return left
    }

    /// Count occurrences of target in sorted array
    /// - Complexity: O(log n) time, O(1) space
    public static func countOccurrences<T: Comparable>(_ array: [T], target: T) -> Int {
        let lower = lowerBound(array, target: target)

        if lower == array.count || array[lower] != target {
            return 0
        }

        let upper = upperBound(array, target: target)
        return upper - lower
    }

    // MARK: - Linear Search Variants

    /// Standard linear search
    /// - Complexity: O(n) time, O(1) space
    public static func linearSearch<T: Equatable>(_ array: [T], target: T) -> Int? {
        for (index, element) in array.enumerated() {
            if element == target {
                return index
            }
        }
        return nil
    }

    /// Find all occurrences of target
    /// - Complexity: O(n) time, O(k) space where k is number of occurrences
    public static func findAll<T: Equatable>(_ array: [T], target: T) -> [Int] {
        return array.enumerated().compactMap { index, element in
            element == target ? index : nil
        }
    }

    /// Sentinel linear search (faster for arrays that can be modified)
    /// - Complexity: O(n) time, O(1) space
    public static func sentinelSearch<T: Equatable>(_ array: inout [T], target: T) -> Int? {
        let lastElement = array.last
        array[array.count - 1] = target

        var i = 0
        while array[i] != target {
            i += 1
        }

        // Restore original value
        if let last = lastElement {
            array[array.count - 1] = last
        }

        if i < array.count - 1 || lastElement == target {
            return i
        }

        return nil
    }

    // MARK: - Advanced Search Algorithms

    /// Jump search for sorted arrays
    /// - Complexity: O(√n) time, O(1) space
    public static func jumpSearch<T: Comparable>(_ array: [T], target: T) -> Int? {
        let n = array.count
        let step = Int(sqrt(Double(n)))

        var prev = 0
        var current = 0

        // Jump to find block containing target
        while current < n && array[min(step, n) - 1] < target {
            prev = current
            current += step
            if prev >= n {
                return nil
            }
        }

        // Linear search in block
        while array[prev] < target {
            prev += 1
            if prev == min(current, n) {
                return nil
            }
        }

        if array[prev] == target {
            return prev
        }

        return nil
    }

    /// Interpolation search for uniformly distributed sorted arrays
    /// - Complexity: O(log log n) average, O(n) worst case
    public static func interpolationSearch(_ array: [Int], target: Int) -> Int? {
        var low = 0
        var high = array.count - 1

        while low <= high && target >= array[low] && target <= array[high] {
            if low == high {
                return array[low] == target ? low : nil
            }

            // Interpolation formula
            let pos = low + ((target - array[low]) * (high - low)) / (array[high] - array[low])

            if array[pos] == target {
                return pos
            } else if array[pos] < target {
                low = pos + 1
            } else {
                high = pos - 1
            }
        }

        return nil
    }

    /// Exponential search for unbounded arrays
    /// - Complexity: O(log n) time, O(1) space
    public static func exponentialSearch<T: Comparable>(_ array: [T], target: T) -> Int? {
        if array.isEmpty {
            return nil
        }

        if array[0] == target {
            return 0
        }

        // Find range for binary search
        var bound = 1
        while bound < array.count && array[bound] <= target {
            bound *= 2
        }

        // Binary search in found range
        let left = bound / 2
        let right = min(bound, array.count - 1)
        let subArray = Array(array[left...right])

        if let index = binarySearch(subArray, target: target) {
            return left + index
        }

        return nil
    }

    /// Ternary search for sorted arrays
    /// - Complexity: O(log₃ n) time, O(1) space
    public static func ternarySearch<T: Comparable>(_ array: [T], target: T) -> Int? {
        var left = 0
        var right = array.count - 1

        while left <= right {
            let mid1 = left + (right - left) / 3
            let mid2 = right - (right - left) / 3

            if array[mid1] == target {
                return mid1
            }
            if array[mid2] == target {
                return mid2
            }

            if target < array[mid1] {
                right = mid1 - 1
            } else if target > array[mid2] {
                left = mid2 + 1
            } else {
                left = mid1 + 1
                right = mid2 - 1
            }
        }

        return nil
    }

    /// Fibonacci search for sorted arrays
    /// - Complexity: O(log n) time, O(1) space
    public static func fibonacciSearch<T: Comparable>(_ array: [T], target: T) -> Int? {
        let n = array.count

        // Initialize Fibonacci numbers
        var fib2 = 0  // (m-2)th Fibonacci number
        var fib1 = 1  // (m-1)th Fibonacci number
        var fibM = fib2 + fib1  // mth Fibonacci number

        // Find smallest Fibonacci number >= n
        while fibM < n {
            fib2 = fib1
            fib1 = fibM
            fibM = fib2 + fib1
        }

        var offset = -1

        while fibM > 1 {
            let i = min(offset + fib2, n - 1)

            if array[i] < target {
                fibM = fib1
                fib1 = fib2
                fib2 = fibM - fib1
                offset = i
            } else if array[i] > target {
                fibM = fib2
                fib1 = fib1 - fib2
                fib2 = fibM - fib1
            } else {
                return i
            }
        }

        // Check last element
        if fib1 == 1 && offset + 1 < n && array[offset + 1] == target {
            return offset + 1
        }

        return nil
    }

    // MARK: - Specialized Search Algorithms

    /// Search in rotated sorted array
    /// - Complexity: O(log n) time, O(1) space
    public static func searchRotated<T: Comparable>(_ array: [T], target: T) -> Int? {
        var left = 0
        var right = array.count - 1

        while left <= right {
            let mid = left + (right - left) / 2

            if array[mid] == target {
                return mid
            }

            // Determine which half is sorted
            if array[left] <= array[mid] {
                // Left half is sorted
                if array[left] <= target && target < array[mid] {
                    right = mid - 1
                } else {
                    left = mid + 1
                }
            } else {
                // Right half is sorted
                if array[mid] < target && target <= array[right] {
                    left = mid + 1
                } else {
                    right = mid - 1
                }
            }
        }

        return nil
    }

    /// Find peak element in array (local maximum)
    /// - Complexity: O(log n) time, O(1) space
    public static func findPeak<T: Comparable>(_ array: [T]) -> Int? {
        guard !array.isEmpty else { return nil }

        var left = 0
        var right = array.count - 1

        while left < right {
            let mid = left + (right - left) / 2

            if array[mid] > array[mid + 1] {
                right = mid
            } else {
                left = mid + 1
            }
        }

        return left
    }

    /// Two-pointer search for pair with given sum
    /// - Complexity: O(n) time for sorted array, O(1) space
    public static func twoSum(_ array: [Int], target: Int) -> (Int, Int)? {
        var left = 0
        var right = array.count - 1

        while left < right {
            let sum = array[left] + array[right]

            if sum == target {
                return (left, right)
            } else if sum < target {
                left += 1
            } else {
                right -= 1
            }
        }

        return nil
    }

    /// Three-pointer search for triplet with given sum
    /// - Complexity: O(n²) time, O(1) space
    public static func threeSum(_ array: [Int], target: Int) -> [(Int, Int, Int)] {
        var result: [(Int, Int, Int)] = []
        let sortedArray = array.sorted()

        for i in 0..<sortedArray.count - 2 {
            // Skip duplicates
            if i > 0 && sortedArray[i] == sortedArray[i - 1] {
                continue
            }

            var left = i + 1
            var right = sortedArray.count - 1

            while left < right {
                let sum = sortedArray[i] + sortedArray[left] + sortedArray[right]

                if sum == target {
                    result.append((i, left, right))

                    // Skip duplicates
                    while left < right && sortedArray[left] == sortedArray[left + 1] {
                        left += 1
                    }
                    while left < right && sortedArray[right] == sortedArray[right - 1] {
                        right -= 1
                    }

                    left += 1
                    right -= 1
                } else if sum < target {
                    left += 1
                } else {
                    right -= 1
                }
            }
        }

        return result
    }

    // MARK: - K-th Element Algorithms

    /// Quick select to find k-th smallest element
    /// - Complexity: O(n) average, O(n²) worst case
    public static func quickSelect<T: Comparable>(_ array: inout [T], k: Int) -> T? {
        guard k > 0 && k <= array.count else { return nil }
        return quickSelectHelper(&array, k - 1, 0, array.count - 1)
    }

    private static func quickSelectHelper<T: Comparable>(
        _ array: inout [T],
        _ k: Int,
        _ left: Int,
        _ right: Int
    ) -> T? {
        if left == right {
            return array[left]
        }

        let pivotIndex = partition(&array, left, right)

        if k == pivotIndex {
            return array[k]
        } else if k < pivotIndex {
            return quickSelectHelper(&array, k, left, pivotIndex - 1)
        } else {
            return quickSelectHelper(&array, k, pivotIndex + 1, right)
        }
    }

    private static func partition<T: Comparable>(_ array: inout [T], _ left: Int, _ right: Int) -> Int {
        let pivot = array[right]
        var i = left

        for j in left..<right {
            if array[j] <= pivot {
                array.swapAt(i, j)
                i += 1
            }
        }

        array.swapAt(i, right)
        return i
    }

    // MARK: - Pattern Searching

    /// Naive pattern search in string
    /// - Complexity: O(nm) time where n is text length, m is pattern length
    public static func naivePatternSearch(_ text: String, pattern: String) -> [Int] {
        var result: [Int] = []
        let textArray = Array(text)
        let patternArray = Array(pattern)
        let n = textArray.count
        let m = patternArray.count

        for i in 0...(n - m) {
            var j = 0

            while j < m && textArray[i + j] == patternArray[j] {
                j += 1
            }

            if j == m {
                result.append(i)
            }
        }

        return result
    }

    // MARK: - Parallel Search (iOS Optimized)

    /// Parallel search using GCD for large datasets
    @available(iOS 13.0, macOS 10.15, *)
    public static func parallelSearch<T: Equatable>(
        _ array: [T],
        target: T
    ) async -> [Int] {
        let chunkSize = max(1, array.count / ProcessInfo.processInfo.activeProcessorCount)

        return await withTaskGroup(of: [Int].self) { group in
            for (offset, chunk) in array.chunked(into: chunkSize).enumerated() {
                group.addTask {
                    var indices: [Int] = []
                    for (index, element) in chunk.enumerated() {
                        if element == target {
                            indices.append(offset * chunkSize + index)
                        }
                    }
                    return indices
                }
            }

            var allIndices: [Int] = []
            for await indices in group {
                allIndices.append(contentsOf: indices)
            }

            return allIndices.sorted()
        }
    }
}

// MARK: - Helper Extensions

private extension Array {
    /// Split array into chunks of specified size
    func chunked(into size: Int) -> [[Element]] {
        return stride(from: 0, to: count, by: size).map {
            Array(self[$0..<Swift.min($0 + size, count)])
        }
    }
}

// MARK: - Search Result Type

/// Result type for detailed search operations
public struct SearchResult<T> {
    public let index: Int?
    public let element: T?
    public let comparisons: Int
    public let timeElapsed: TimeInterval

    public var found: Bool {
        return index != nil
    }
}

// MARK: - Performance Measurement

/// Utility for measuring search performance
public struct SearchBenchmark {
    /// Measure search algorithm performance
    public static func measure<T: Comparable>(
        _ array: [T],
        target: T,
        algorithm: (([T], T) -> Int?)
    ) -> SearchResult<T> {
        let startTime = CFAbsoluteTimeGetCurrent()
        var comparisons = 0

        // Hook for counting comparisons (simplified for example)
        let index = algorithm(array, target)

        let timeElapsed = CFAbsoluteTimeGetCurrent() - startTime

        return SearchResult(
            index: index,
            element: index.map { array[$0] },
            comparisons: comparisons,
            timeElapsed: timeElapsed
        )
    }

    /// Compare multiple search algorithms
    public static func compare<T: Comparable>(
        _ array: [T],
        target: T,
        algorithms: [(name: String, function: ([T], T) -> Int?)]
    ) -> [(name: String, result: SearchResult<T>)] {
        return algorithms.map { (name, function) in
            (name, measure(array, target: target, algorithm: function))
        }
    }
}

File: c/src/searching.c
/**
 * @file searching.c
 * @brief Implementation of searching algorithms
 */

#include "searching.h"
#include <stdlib.h>
#include <string.h>
#include <math.h>

/* Linear Search */
search_result_t linear_search(const int* arr, size_t n, int target) {
    search_result_t result = {false, 0, 0};

    if (!arr || n == 0) return result;

    for (size_t i = 0; i < n; i++) {
        result.comparisons++;
        if (arr[i] == target) {
            result.found = true;
            result.index = i;
            break;
        }
    }

    return result;
}

/* Generic Linear Search */
search_result_t linear_search_generic(const void* arr, size_t n, size_t elem_size,
                                      const void* target,
                                      int (*compare)(const void*, const void*)) {
    search_result_t result = {false, 0, 0};

    if (!arr || n == 0 || !target || !compare) return result;

    const char* array = (const char*)arr;

    for (size_t i = 0; i < n; i++) {
        result.comparisons++;
        if (compare(array + i * elem_size, target) == 0) {
            result.found = true;
            result.index = i;
            break;
        }
    }

    return result;
}

/* Binary Search (iterative) */
search_result_t binary_search(const int* arr, size_t n, int target) {
    search_result_t result = {false, 0, 0};

    if (!arr || n == 0) return result;

    size_t left = 0;
    size_t right = n - 1;

    while (left <= right) {
        size_t mid = left + (right - left) / 2;
        result.comparisons++;

        if (arr[mid] == target) {
            result.found = true;
            result.index = mid;
            break;
        } else if (arr[mid] < target) {
            left = mid + 1;
        } else {
            if (mid == 0) break;  /* Prevent underflow */
            right = mid - 1;
        }
    }

    return result;
}

/* Binary Search (recursive helper) */
static search_result_t binary_search_recursive_helper(const int* arr, size_t left,
                                                      size_t right, int target,
                                                      size_t comparisons) {
    search_result_t result = {false, 0, comparisons};

    if (left > right) return result;

    size_t mid = left + (right - left) / 2;
    result.comparisons++;

    if (arr[mid] == target) {
        result.found = true;
        result.index = mid;
    } else if (arr[mid] < target) {
        return binary_search_recursive_helper(arr, mid + 1, right, target, result.comparisons);
    } else {
        if (mid == 0) return result;
        return binary_search_recursive_helper(arr, left, mid - 1, target, result.comparisons);
    }

    return result;
}

/* Binary Search (recursive) */
search_result_t binary_search_recursive(const int* arr, size_t n, int target) {
    if (!arr || n == 0) {
        search_result_t result = {false, 0, 0};
        return result;
    }

    return binary_search_recursive_helper(arr, 0, n - 1, target, 0);
}

/* Generic Binary Search */
search_result_t binary_search_generic(const void* arr, size_t n, size_t elem_size,
                                      const void* target,
                                      int (*compare)(const void*, const void*)) {
    search_result_t result = {false, 0, 0};

    if (!arr || n == 0 || !target || !compare) return result;

    const char* array = (const char*)arr;
    size_t left = 0;
    size_t right = n - 1;

    while (left <= right) {
        size_t mid = left + (right - left) / 2;
        result.comparisons++;

        int cmp = compare(array + mid * elem_size, target);

        if (cmp == 0) {
            result.found = true;
            result.index = mid;
            break;
        } else if (cmp < 0) {
            left = mid + 1;
        } else {
            if (mid == 0) break;
            right = mid - 1;
        }
    }

    return result;
}

/* Jump Search */
search_result_t jump_search(const int* arr, size_t n, int target) {
    search_result_t result = {false, 0, 0};

    if (!arr || n == 0) return result;

    size_t step = (size_t)sqrt(n);
    size_t prev = 0;

    /* Jump to find the block */
    while (arr[prev] < target) {
        result.comparisons++;
        size_t next = prev + step;

        if (next >= n) {
            next = n - 1;
        }

        if (arr[next] >= target) {
            break;
        }

        prev = next;
        if (prev >= n - 1) break;
    }

    /* Linear search in the block */
    for (size_t i = prev; i < n && i < prev + step; i++) {
        result.comparisons++;
        if (arr[i] == target) {
            result.found = true;
            result.index = i;
            break;
        }
        if (arr[i] > target) {
            break;
        }
    }

    return result;
}

/* Interpolation Search */
search_result_t interpolation_search(const int* arr, size_t n, int target) {
    search_result_t result = {false, 0, 0};

    if (!arr || n == 0) return result;

    size_t low = 0;
    size_t high = n - 1;

    while (low <= high && target >= arr[low] && target <= arr[high]) {
        result.comparisons++;

        if (low == high) {
            if (arr[low] == target) {
                result.found = true;
                result.index = low;
            }
            break;
        }

        /* Calculate position using interpolation formula */
        size_t pos = low + (((double)(high - low) / (arr[high] - arr[low])) *
                           (target - arr[low]));

        if (arr[pos] == target) {
            result.found = true;
            result.index = pos;
            break;
        } else if (arr[pos] < target) {
            low = pos + 1;
        } else {
            if (pos == 0) break;
            high = pos - 1;
        }
    }

    return result;
}

/* Exponential Search */
search_result_t exponential_search(const int* arr, size_t n, int target) {
    search_result_t result = {false, 0, 0};

    if (!arr || n == 0) return result;

    /* If target is at first position */
    result.comparisons++;
    if (arr[0] == target) {
        result.found = true;
        result.index = 0;
        return result;
    }

    /* Find range for binary search */
    size_t i = 1;
    while (i < n && arr[i] <= target) {
        result.comparisons++;
        if (arr[i] == target) {
            result.found = true;
            result.index = i;
            return result;
        }
        i *= 2;
    }

    /* Apply binary search on the range */
    size_t left = i / 2;
    size_t right = (i < n) ? i : n - 1;

    while (left <= right) {
        size_t mid = left + (right - left) / 2;
        result.comparisons++;

        if (arr[mid] == target) {
            result.found = true;
            result.index = mid;
            break;
        } else if (arr[mid] < target) {
            left = mid + 1;
        } else {
            if (mid == 0) break;
            right = mid - 1;
        }
    }

    return result;
}

/* Fibonacci Search */
search_result_t fibonacci_search(const int* arr, size_t n, int target) {
    search_result_t result = {false, 0, 0};

    if (!arr || n == 0) return result;

    /* Initialize Fibonacci numbers */
    size_t fib2 = 0;  /* (m-2)'th Fibonacci number */
    size_t fib1 = 1;  /* (m-1)'th Fibonacci number */
    size_t fib = fib2 + fib1;  /* m'th Fibonacci number */

    /* Find the smallest Fibonacci number >= n */
    while (fib < n) {
        fib2 = fib1;
        fib1 = fib;
        fib = fib2 + fib1;
    }

    size_t offset = 0;

    while (fib > 1) {
        /* Check if fib2 is valid index */
        size_t i = (offset + fib2 < n - 1) ? offset + fib2 : n - 1;

        result.comparisons++;

        if (arr[i] < target) {
            fib = fib1;
            fib1 = fib2;
            fib2 = fib - fib1;
            offset = i;
        } else if (arr[i] > target) {
            fib = fib2;
            fib1 = fib1 - fib2;
            fib2 = fib - fib1;
        } else {
            result.found = true;
            result.index = i;
            return result;
        }
    }

    /* Check last element */
    if (offset < n) {
        result.comparisons++;
        if (arr[offset] == target) {
            result.found = true;
            result.index = offset;
        }
    }

    return result;
}

/* Ternary Search */
search_result_t ternary_search(const int* arr, size_t n, int target) {
    search_result_t result = {false, 0, 0};

    if (!arr || n == 0) return result;

    size_t left = 0;
    size_t right = n - 1;

    while (left <= right) {
        size_t mid1 = left + (right - left) / 3;
        size_t mid2 = right - (right - left) / 3;

        result.comparisons++;
        if (arr[mid1] == target) {
            result.found = true;
            result.index = mid1;
            return result;
        }

        result.comparisons++;
        if (arr[mid2] == target) {
            result.found = true;
            result.index = mid2;
            return result;
        }

        if (target < arr[mid1]) {
            if (mid1 == 0) break;
            right = mid1 - 1;
        } else if (target > arr[mid2]) {
            left = mid2 + 1;
        } else {
            left = mid1 + 1;
            if (mid2 == 0) break;
            right = mid2 - 1;
        }
    }

    return result;
}

/* Find minimum element */
int find_min(const int* arr, size_t n) {
    if (!arr || n == 0) return 0;

    int min = arr[0];
    for (size_t i = 1; i < n; i++) {
        if (arr[i] < min) {
            min = arr[i];
        }
    }
    return min;
}

/* Find maximum element */
int find_max(const int* arr, size_t n) {
    if (!arr || n == 0) return 0;

    int max = arr[0];
    for (size_t i = 1; i < n; i++) {
        if (arr[i] > max) {
            max = arr[i];
        }
    }
    return max;
}

/* Quick select algorithm for kth element */
static int partition_quickselect(int* arr, size_t low, size_t high) {
    int pivot = arr[high];
    size_t i = low;

    for (size_t j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
            i++;
        }
    }

    int temp = arr[i];
    arr[i] = arr[high];
    arr[high] = temp;

    return i;
}

static int quickselect(int* arr, size_t low, size_t high, size_t k) {
    if (low == high) {
        return arr[low];
    }

    size_t pivot_index = partition_quickselect(arr, low, high);

    if (k == pivot_index) {
        return arr[k];
    } else if (k < pivot_index) {
        return quickselect(arr, low, pivot_index - 1, k);
    } else {
        return quickselect(arr, pivot_index + 1, high, k);
    }
}

/* Find kth smallest element */
int find_kth_smallest(int* arr, size_t n, size_t k) {
    if (!arr || n == 0 || k == 0 || k > n) return 0;

    /* Create a copy to avoid modifying original array */
    int* temp = malloc(n * sizeof(int));
    memcpy(temp, arr, n * sizeof(int));

    int result = quickselect(temp, 0, n - 1, k - 1);

    free(temp);
    return result;
}

/* Find kth largest element */
int find_kth_largest(int* arr, size_t n, size_t k) {
    if (!arr || n == 0 || k == 0 || k > n) return 0;

    /* kth largest is (n - k + 1)th smallest */
    return find_kth_smallest(arr, n, n - k + 1);
}

/* Find median */
int find_median(int* arr, size_t n) {
    if (!arr || n == 0) return 0;

    if (n % 2 == 1) {
        /* Odd number of elements */
        return find_kth_smallest(arr, n, (n + 1) / 2);
    } else {
        /* Even number of elements - return lower median */
        return find_kth_smallest(arr, n, n / 2);
    }
}

/* Lower bound - first element >= value */
size_t lower_bound(const int* arr, size_t n, int value) {
    if (!arr || n == 0) return 0;

    size_t left = 0;
    size_t right = n;

    while (left < right) {
        size_t mid = left + (right - left) / 2;

        if (arr[mid] < value) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }

    return left;
}

/* Upper bound - first element > value */
size_t upper_bound(const int* arr, size_t n, int value) {
    if (!arr || n == 0) return 0;

    size_t left = 0;
    size_t right = n;

    while (left < right) {
        size_t mid = left + (right - left) / 2;

        if (arr[mid] <= value) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }

    return left;
}

/* Equal range - find range of elements equal to value */
void equal_range(const int* arr, size_t n, int value, size_t* lower, size_t* upper) {
    if (!arr || n == 0 || !lower || !upper) return;

    *lower = lower_bound(arr, n, value);
    *upper = upper_bound(arr, n, value);
}

/* Count occurrences of a value */
size_t count_occurrences(const int* arr, size_t n, int value) {
    if (!arr || n == 0) return 0;

    size_t lower, upper;
    equal_range(arr, n, value, &lower, &upper);

    return upper - lower;
}

/* Find peak element in 1D array */
size_t find_peak_1d(const int* arr, size_t n) {
    if (!arr || n == 0) return 0;
    if (n == 1) return 0;

    size_t left = 0;
    size_t right = n - 1;

    while (left <= right) {
        size_t mid = left + (right - left) / 2;

        /* Check if mid is a peak */
        bool left_ok = (mid == 0 || arr[mid] >= arr[mid - 1]);
        bool right_ok = (mid == n - 1 || arr[mid] >= arr[mid + 1]);

        if (left_ok && right_ok) {
            return mid;
        }

        /* Move to the side with larger neighbor */
        if (mid > 0 && arr[mid - 1] > arr[mid]) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    return 0;
}

/* Find peak element in 2D matrix */
void find_peak_2d(const int** matrix, size_t rows, size_t cols,
                  size_t* peak_row, size_t* peak_col) {
    if (!matrix || rows == 0 || cols == 0 || !peak_row || !peak_col) return;

    *peak_row = 0;
    *peak_col = 0;

    size_t left = 0;
    size_t right = cols - 1;

    while (left <= right) {
        size_t mid_col = left + (right - left) / 2;

        /* Find maximum in mid column */
        size_t max_row = 0;
        for (size_t i = 1; i < rows; i++) {
            if (matrix[i][mid_col] > matrix[max_row][mid_col]) {
                max_row = i;
            }
        }

        /* Check if it's a peak */
        bool left_ok = (mid_col == 0 ||
                        matrix[max_row][mid_col] >= matrix[max_row][mid_col - 1]);
        bool right_ok = (mid_col == cols - 1 ||
                         matrix[max_row][mid_col] >= matrix[max_row][mid_col + 1]);

        if (left_ok && right_ok) {
            *peak_row = max_row;
            *peak_col = mid_col;
            return;
        }

        /* Move to the side with larger neighbor */
        if (mid_col > 0 && matrix[max_row][mid_col - 1] > matrix[max_row][mid_col]) {
            right = mid_col - 1;
        } else {
            left = mid_col + 1;
        }
    }
}

/* Find majority element (Boyer-Moore Voting Algorithm) */
int find_majority_element(const int* arr, size_t n, bool* has_majority) {
    if (!arr || n == 0 || !has_majority) {
        if (has_majority) *has_majority = false;
        return 0;
    }

    /* Find candidate */
    int candidate = arr[0];
    int count = 1;

    for (size_t i = 1; i < n; i++) {
        if (arr[i] == candidate) {
            count++;
        } else {
            count--;
            if (count == 0) {
                candidate = arr[i];
                count = 1;
            }
        }
    }

    /* Verify if candidate is majority */
    count = 0;
    for (size_t i = 0; i < n; i++) {
        if (arr[i] == candidate) {
            count++;
        }
    }

    *has_majority = (count > n / 2);
    return candidate;
}

/* Two Sum - find two indices that sum to target */
bool two_sum(const int* arr, size_t n, int target, size_t* idx1, size_t* idx2) {
    if (!arr || n < 2 || !idx1 || !idx2) return false;

    /* Create array of indices for sorting */
    size_t* indices = malloc(n * sizeof(size_t));
    int* sorted = malloc(n * sizeof(int));

    for (size_t i = 0; i < n; i++) {
        indices[i] = i;
        sorted[i] = arr[i];
    }

    /* Sort array and indices together */
    for (size_t i = 0; i < n - 1; i++) {
        for (size_t j = i + 1; j < n; j++) {
            if (sorted[i] > sorted[j]) {
                int temp = sorted[i];
                sorted[i] = sorted[j];
                sorted[j] = temp;

                size_t temp_idx = indices[i];
                indices[i] = indices[j];
                indices[j] = temp_idx;
            }
        }
    }

    /* Two pointer approach */
    size_t left = 0;
    size_t right = n - 1;
    bool found = false;

    while (left < right) {
        int sum = sorted[left] + sorted[right];

        if (sum == target) {
            *idx1 = indices[left];
            *idx2 = indices[right];
            found = true;
            break;
        } else if (sum < target) {
            left++;
        } else {
            right--;
        }
    }

    free(indices);
    free(sorted);
    return found;
}

/* Three Sum - find three indices that sum to target */
bool three_sum(const int* arr, size_t n, int target,
                size_t* idx1, size_t* idx2, size_t* idx3) {
    if (!arr || n < 3 || !idx1 || !idx2 || !idx3) return false;

    /* Try each element as first element */
    for (size_t i = 0; i < n - 2; i++) {
        /* Create reduced array without element i */
        int* reduced = malloc((n - 1) * sizeof(int));
        size_t* original_indices = malloc((n - 1) * sizeof(size_t));

        size_t k = 0;
        for (size_t j = 0; j < n; j++) {
            if (j != i) {
                reduced[k] = arr[j];
                original_indices[k] = j;
                k++;
            }
        }

        /* Find two sum in reduced array */
        size_t sub_idx1, sub_idx2;
        if (two_sum(reduced, n - 1, target - arr[i], &sub_idx1, &sub_idx2)) {
            *idx1 = i;
            *idx2 = original_indices[sub_idx1];
            *idx3 = original_indices[sub_idx2];

            free(reduced);
            free(original_indices);
            return true;
        }

        free(reduced);
        free(original_indices);
    }

    return false;
}